cpp/src/arrow/compute/memory_resources.cc

```diff
@@ -0,0 +1,307 @@
+ public:
+  explicit CPUDataHolder(const std::shared_ptr<RecordBatch>& record_batch)
+      : DataHolder(MemoryLevel::kCPULevel), record_batch_(std::move(record_batch)) {}
+
+  Result<ExecBatch> Get() override { return ExecBatch(*record_batch_); }
+
+ private:
+  std::shared_ptr<RecordBatch> record_batch_;
+};
+
+namespace {
+
+std::string RandomString(std::size_t length) {
+  const std::string characters =
+      "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
+  std::random_device random_device;
+  std::mt19937 generator(random_device());
+  std::uniform_int_distribution<> distribution(0, characters.size() - 1);
+  std::string random_string;
+  for (std::size_t i = 0; i < length; ++i) {
+    random_string += characters[distribution(generator)];
+  }
+  return random_string;
+}
+
+}  // namespace
+
+Status StoreRecordBatch(const std::shared_ptr<RecordBatch>& record_batch,
+                        const std::shared_ptr<fs::FileSystem>& filesystem,
+                        const std::string& file_path) {
+  auto output = filesystem->OpenOutputStream(file_path).ValueOrDie();
+  auto writer =
+      arrow::ipc::MakeFileWriter(output.get(), record_batch->schema()).ValueOrDie();
+  ARROW_RETURN_NOT_OK(writer->WriteRecordBatch(*record_batch));
+  return writer->Close();
+}
+Result<std::shared_ptr<RecordBatch>> RecoverRecordBatch(
+    const std::shared_ptr<fs::FileSystem>& filesystem, const std::string& file_path) {
+  ARROW_ASSIGN_OR_RAISE(auto input, filesystem->OpenInputFile(file_path));
+  ARROW_ASSIGN_OR_RAISE(auto reader, arrow::ipc::feather::Reader::Open(input));
+  std::shared_ptr<Table> table;
+  ARROW_RETURN_NOT_OK(reader->Read(&table));
+  TableBatchReader batch_iter(*table);
+  ARROW_ASSIGN_OR_RAISE(auto batch, batch_iter.Next());
+  return batch;
+}
+
+class DiskDataHolder : public DataHolder {
+ public:
+  DiskDataHolder(const std::shared_ptr<RecordBatch>& record_batch,
+                 MemoryPool* memory_pool)
+      : DataHolder(MemoryLevel::kDiskLevel), memory_pool_(memory_pool) {
+    std::string root_path;
+    std::string file_name = "data-holder-temp-" + RandomString(64) + ".feather";
+
+    filesystem_ =
+        arrow::fs::FileSystemFromUri(cache_storage_root_path, &root_path).ValueOrDie();
+
+    file_path_ = root_path + file_name;
+    status_ = StoreRecordBatch(record_batch, filesystem_, file_path_);
+  }
+
+  Result<ExecBatch> Get() override {
+    ARROW_RETURN_NOT_OK(status_);
+    ARROW_ASSIGN_OR_RAISE(auto record_batch, RecoverRecordBatch(filesystem_, file_path_));
+    return ExecBatch(*record_batch);
+  }
+
+ private:
+  std::string file_path_;
+  Status status_;
+  MemoryPool* memory_pool_;
+  std::shared_ptr<arrow::fs::FileSystem> filesystem_;
+  const std::string cache_storage_root_path = "file:///tmp/";
+};
+
+class MemoryResources::MemoryResourcesImpl {
+ public:
+  Status AddMemoryResource(std::unique_ptr<MemoryResource> resource) {
+    std::lock_guard<std::mutex> mutation_guard(lock_);
```

Comment:
This might be a bit of a premature optimization.  I don't think `AddMemoryResource` is going to be called a lot.  It might be simpler to do away with the locking (and the mutex) and then you could even get rid of the pimpl at that point.

Reply:
Agree